pkg/operators/ebpf/snapshotter.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/link"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/netnsenter"
+	bpfiterns "github.com/inspektor-gadget/inspektor-gadget/pkg/utils/bpf-iter-ns"
+)
+
+type linkSnapshotter struct {
+	link *link.Iter
+	typ  string
+}
+
+type Snapshotter struct {
+	metadatav1.Snapshotter
+
+	ds       datasource.DataSource
+	accessor datasource.FieldAccessor
+}
+
+func (i *ebpfInstance) populateSnapshotter(t btf.Type, varName string) error {
+	i.logger.Debugf("populating snapshotter %q", varName)
+
+	parts := strings.Split(varName, typeSplitter)
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid snapshotter info: %q", varName)
+	}
+
+	name := parts[0]
+	mapType := parts[1]
+
+	if _, ok := i.snapshotters[name]; ok {
+		i.logger.Debugf("snapshotter %q already defined, skipping", name)
+		return nil
+	}
+
+	var btfStruct *btf.Struct
+	if err := i.collectionSpec.Types.TypeByName(mapType, &btfStruct); err != nil {
+		return fmt.Errorf("finding struct %q in eBPF object: %w", mapType, err)
+	}
+
+	i.logger.Debugf("adding snapshotter %q", name)
+	i.snapshotters[name] = &Snapshotter{
+		Snapshotter: metadatav1.Snapshotter{
+			StructName: btfStruct.Name,
+		},
+	}
+
+	err := i.populateStructDirect(btfStruct)
+	if err != nil {
+		return fmt.Errorf("populating struct %q for snapshotter %q: %w", btfStruct.Name, name, err)
+	}
+
+	return nil
+}
+
+func (i *ebpfInstance) runSnapshotters() error {
+	for _, l := range i.linksSnapshotters {
+		i.logger.Debugf("starting snapshotter")
+		switch l.typ {
+		case "task":
+			buf, err := bpfiterns.Read(l.link)
+			if err != nil {
+				return fmt.Errorf("reading iterator: %w", err)
+			}
+			// TODO: we need a link from iter to map
```

Comment:
Are you already working on this? If you're not, I could do it in a follow-up PR.

Reply:
No, I'm not. I think this needs changes to our gadget macros - I didn't want to touch any of that for this PR to be as compatible as possible.

Reply:
Draft PR: https://github.com/inspektor-gadget/inspektor-gadget/pull/2637